File: problem-solving/baekjoon/step18/1931_회의실배정.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscanln(reader, &n)
	var times = make([][]int, n)
	for i := 0; i < n; i++ {
		var start, end int
		fmt.Fscanln(reader, &start, &end)
		times[i] = []int{start, end}
	}
	maxCount := greedyAlgorithm(times, n)
	fmt.Println(maxCount)
}

func greedyAlgorithm(times [][]int, n int) (maxCount int) {
	sort.Slice(times, func(i, j int) bool {
		if times[j][1] > times[i][1] {
			return true
		} else if times[j][1] == times[i][1] {
			return times[j][0] > times[i][0]
		}
		return false
	})
	previousTime := []int{}
	for i := 0; i < n; i++ {
		if i == 0 {
			maxCount++
			previousTime = times[i]
			continue
		}
		if times[i][0] >= previousTime[1] {
			maxCount++
			previousTime = times[i]
		}
	}
	return
}

File: problem-solving/baekjoon/step18/1931_회의실배정.py
import sys

def greedy_algorithm(times, n):
    meeting_count = 0
    times.sort(key=lambda x: (x[1], x[0]))
    previous_time = []
    for i in range(0, n):
        if i == 0:
            meeting_count += 1
            previous_time = times[i]
            continue
        if times[i][0] >= previous_time[1]:
            meeting_count += 1
            previous_time = times[i]
    return meeting_count

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    times = []
    for i in range(0, n):
        start, end = map(int, sys.stdin.readline().split())
        times.append([start,end])
        
    max_count = greedy_algorithm(times, n)
    print(max_count)